packages/editor/src/store/actions.js

```diff
@@ -32,6 +33,120 @@ import {
 	getNotificationArgumentsForSaveFail,
 	getNotificationArgumentsForTrashFail,
 } from './utils/notice-builder';
+import { awaitNextStateChange, getRegistry } from './controls';
+import * as sources from './block-sources';
+
+/**
+ * Map of Registry instance to WeakMap of dependencies by custom source.
+ *
+ * @type WeakMap<WPDataRegistry,WeakMap<WPBlockAttributeSource,Object>>
+ */
+const lastBlockSourceDependenciesByRegistry = new WeakMap;
+
+/**
+ * Given a blocks array, returns a blocks array with sourced attribute values
+ * applied. The reference will remain consistent with the original argument if
+ * no attribute values must be overridden. If sourced values are applied, the
+ * return value will be a modified copy of the original array.
+ *
+ * @param {WPBlock[]} blocks Original blocks array.
+ *
+ * @return {WPBlock[]} Blocks array with sourced values applied.
+ */
+function* getBlocksWithSourcedAttributes( blocks ) {
+	const registry = yield getRegistry();
+
+	let workingBlocks = blocks;
+	for ( let i = 0; i < blocks.length; i++ ) {
+		let block = blocks[ i ];
+		const blockType = yield select( 'core/blocks', 'getBlockType', block.name );
+
+		for ( const [ attributeName, schema ] of Object.entries( blockType.attributes ) ) {
+			if ( ! sources[ schema.source ] || ! sources[ schema.source ].apply ) {
+				continue;
+			}
+
+			if ( ! lastBlockSourceDependenciesByRegistry.has( registry ) ) {
+				continue;
+			}
```

Comment:
Is the `if` condition expected to evaluate differently across `for` iterations? If not, we could place it as a condition for stepping into the `for` loop.